ARROW-4226: [C++] Add sparse CSF tensor support

rok opened the pull request:
This is to resolve [ARROW-4226](https://issues.apache.org/jira/browse/ARROW-4226).

pitrou commented on format/SparseTensor.fbs:
```diff
@@ -108,9 +108,22 @@ table SparseMatrixIndexCSR {
+  indptrBuffer: Buffer;
+  indicesType: Int;
+  indicesBuffer: Buffer;
+  axisOrder: [long];
```
It would be nice to add a couple comments explaining the format, as done above for `SparseMatrixIndexCSR`.

rok replied:
Will do.

pitrou commented on format/SparseTensor.fbs:
```diff
@@ -108,9 +108,22 @@ table SparseMatrixIndexCSR {
+/// Compressed Sparse Fiber (CSF) sparse tensor format
+///
+/// CSF is a generalization of compressed sparse row (CSR) index.
+/// CSF compresses a tensor into one three one-dimensional tensors.
```
Is there a typo here? "one three one-dimensional tensors"

pitrou commented on cpp/src/arrow/sparse_tensor.cc:
```diff
@@ -279,6 +279,123 @@ class SparseTensorConverter<TYPE, SparseCSRIndex>
+    std::shared_ptr<Tensor> indices_tensor =
+        std::make_shared<Tensor>(index_value_type_, indices_buffer, indices_shape);
+
+    sparse_index = std::make_shared<SparseCSFIndex>(indptr_tensor, indices_tensor, indptr_tensor);
```
`indptr_tensor` is also given as axis order? How is that?

rok replied:
Yeah, that was just to try something out.

rok commented:
This is missing the `MakeSparseTensorFromTensor` part for `CSF`.

mrkn commented:
@rok Thank you for working to implement CSF format.

We should research the existing implementations of CSF format so that our implementation can be used to exchange the data without copying buffers.

Did you find the existing libraries that have CSF format?  If you know some such libraries, could you tell me them for the investigation of the data layout?

rok commented:
> We should research the existing implementations of CSF format so that our implementation can be used to exchange the data without copying buffers.

Agreed! :)

> Did you find the existing libraries that have CSF format? If you know some such libraries, could you tell me them for the investigation of the data layout?

To my knowledge CSF is implemented in [taco](https://github.com/tensor-compiler/taco/search?q=csf&unscoped_q=csf) and in pydata/sparse as `GXCS`, see [this PR](https://github.com/pydata/sparse/pull/258). I did not research the taco implementation.
As for `sparse.GXCS`: I'm skeptical on the way it indices values, [see discussion here](https://github.com/pydata/sparse/issues/302). My proposal uses offsets and it should be easily possible to interface with `sparse.GXCS`.

My main reference for implementation is `Figure 2` in [this paper](http://shaden.io/pub-files/smith2017knl.pdf). I'm not sure this implementation is optimal, please give me your thoughts.

rok commented:
Another point: CSF is a generalization of CSR and CSC. Because of this transformations `CSF <-> CSC` and `CSF <-> CSR` don't require changes to the index. I'm proposing a potential `COO -> CSF` in this PR, so we just need `COO <- CSF` and we have all sparse index transformations covered.
This would enable us to compare SparseTensors with different indices.

mrkn commented:
`GCXS` in pydata/sparse seems an implementation of GCRS and GCCS proposed in [this article](https://ieeexplore.ieee.org/document/7237032).

I confirmed that taco supports CSF format.

Now I'm reading [this paper](https://dl.acm.org/citation.cfm?id=3276493) and the implementation of taco to study taco's abstraction sparse tensor format.

I guess we need more researches on the real-world implementations of CSF format right now.

rok commented:
There is also a CSF implementation by @ShadenSmith in [SPLATT](https://github.com/ShadenSmith/splatt/blob/master/src/csf.c).

ShadenSmith commented:
Hello! It's great to see your interest in the CSF format. I'm happy to answer any questions that you have.

rok commented:
Hi @ShadenSmith. We're currently discussing layout of arrow's CSF implementation. As @mrkn mentioned we want to design it in a way that lets us avoid copying data when interfacing with existing libraries and rather make it possible to just pass pointers.

This PR currently proposes a structure like:
```
table SparseTensorIndexCSF {
  indptrType: Int;
  indptrBuffer: Buffer;
  indptrOffsets: [int];
  indicesType: Int;
  indicesBuffer: Buffer;
  indicesOffsets: [int];
  axisOrder: [long];
}
```
Here `indices` and `inptrs` are stored `indptrBuffer` and `indicesBuffer` respectively. Offsets are used to split these buffers into per dimension `indices` and `indptrs`. [See here for comments](https://github.com/apache/arrow/blob/0a8a06c554e82b3c570a88aeb5bb03683634b114/format/SparseTensor.fbs#L111).

If I understand correctly SPLATT uses a different structure where `indices` and `indptrs` are stored in a '2D' data structure where `depth` accesses data per dimension, eg: `fptr[dimension][fiber]` and `fids[dimension][fiber]`.

@ShadenSmith could you say:
* do you think other CSF implementations are laid out like SPLATTs?
* what number (range) of dimensions would you expect to encounter / need in real world applications?

ShadenSmith commented:
Hi @rok, the other CSF implementations that I have seen either include the code from SPLATT or also just use the 2D layout for the per-mode structures. Here is an [example](https://github.com/isratnisa/B-CSF/blob/master/src/util.h). I'm not sure if taco uses the 1D or 2D implementation and did not find the details from a quick search of the repo.

I think five or six modes (dimensions) is the largest I have encountered in an application, with three being the most common. I have heard of some ongoing research with O(100-1000) mode tensors, but haven't seen results yet and am not sure something so high dimensional will pan out.

rok commented:
> Hi @rok, the other CSF implementations that I have seen either include the code from SPLATT or also just use the 2D layout for the per-mode structures. Here is an [example](https://github.com/isratnisa/B-CSF/blob/master/src/util.h). I'm not sure if taco uses the 1D or 2D implementation and did not find the details from a quick search of the repo.

Got it, thanks a lot! This is good to know. :)
I'll make another proposal with a 2D structure.

> I think five or six modes (dimensions) is the largest I have encountered in an application, with three being the most common. I have heard of some ongoing research with O(100-1000) mode tensors, but haven't seen results yet and am not sure something so high dimensional will pan out.

Interesting, do you think higher dimensional tensors will be used if available?

ShadenSmith commented:
> Interesting, do you think higher dimensional tensors will be used if available?

Let me first say that I can really only speak for the tensor factorization community, and even there I'm more of a scalability researcher than an application researcher.

Today's software can deal with very high dimensional tensors (SPLATT for example just needs a compile-time flag to increase the max dimensionality). I think the issue is more on the side of the data and quality of a tensor factorization with many dimensions; you may just be learning noise unless the data truly has a nice tensor form with many dimensions. It's quite possible that future application areas do use tensors of that form...I just haven't seen it. Off the top of my head, some of the electronic health record tensors have ~5-10 dimensions, but still not 100-1000 AFAIK.

Also, I would usually not recommend using CSF for tensors with more than 5 or 6 dimensions, unless you have a lot of data that has a non-uniform sparsity pattern.  A simple coordinate format would be better unless the non-zeros arrange nicely into fibers. As you increase the number of dimensions in a tensor, the idea of a fiber gets more and more specific, meaning that the savings you get from storing the sparse tensor in CSF disappear because most fibers will only have a few non-zeros in them.

rok commented:
> Today's software can deal with very high dimensional tensors (SPLATT for example just needs a compile-time flag to increase the max dimensionality). I think the issue is more on the side of the data and quality of a tensor factorization with many dimensions; you may just be learning noise unless the data truly has a nice tensor form with many dimensions. It's quite possible that future application areas do use tensors of that form...I just haven't seen it. Off the top of my head, some of the electronic health record tensors have ~5-10 dimensions, but still not 100-1000 AFAIK.
> 
> Also, I would usually not recommend using CSF for tensors with more than 5 or 6 dimensions, unless you have a lot of data that has a non-uniform sparsity pattern. A simple coordinate format would be better unless the non-zeros arrange nicely into fibers. As you increase the number of dimensions in a tensor, the idea of a fiber gets more and more specific, meaning that the savings you get from storing the sparse tensor in CSF disappear because most fibers will only have a few non-zeros in them.

That is great to know, thanks @ShadenSmith! I think we can cover this area with arrow.
If I understand correctly picking dimension order when going from dense tensor to CSF is not a trivial task. Is there a heuristic you would recommend? E.g. in order of smaller to larger dimension in `tensor.shape`?

@mrkn - I've made another proposal with '2D' structure. Could you just check if it makes sense to you please? I would then proceed with serialization and cython wrapper.

ShadenSmith commented:
> If I understand correctly picking dimension order when going from dense tensor to CSF is not a trivial task. Is there a heuristic you would recommend? E.g. in order of smaller to larger dimension in tensor.shape?

Things get a little weird if you're optimizing for parallelism or specific computational kernels, but in the general case that heuristic is what I would recommend (and is what SPLATT uses). The higher the branching factor of the tree, the more compression that CSF achieves. So we sort the dimensions by length and place the shortest at the top to encourage a tree that's thin at the top and wide at the bottom.

rok commented:
> > If I understand correctly picking dimension order when going from dense tensor to CSF is not a trivial task. Is there a heuristic you would recommend? E.g. in order of smaller to larger dimension in tensor.shape?
> 
> Things get a little weird if you're optimizing for parallelism or specific computational kernels, but in the general case that heuristic is what I would recommend (and is what SPLATT uses). The higher the branching factor of the tree, the more compression that CSF achieves. So we sort the dimensions by length and place the shortest at the top to encourage a tree that's thin at the top and wide at the bottom.

Perfect, I've included this heuristic into dense -> sparse conversion now. Thanks @ShadenSmith!

@mrkn @pitrou, could you please review if this proposal makes sense to you?

rok commented on cpp/src/arrow/sparse_tensor.cc:
```diff
@@ -301,6 +303,130 @@ class SparseTensorConverter<TYPE, SparseCSRIndex>
+    std::vector<int64_t> axis_order = internal::ArgSort(tensor_.shape());
+
+    std::vector<TypedBufferBuilder<c_index_value_type>> indptr_buffer_builders(ndim - 1);
+    std::vector<TypedBufferBuilder<c_index_value_type>> indices_buffer_builders(ndim);
```
Should we [reserve](https://arrow.apache.org/docs/cpp/api/memory.html#_CPPv4N5arrow13BufferBuilder7ReserveEK7int64_t) `TypedBufferBuilder` here? We can't really know in advance, maybe we should approximate?

mrkn commented on cpp/src/arrow/sparse_tensor.h:
```diff
@@ -36,7 +36,9 @@ struct SparseTensorFormat {
     /// Compressed sparse row (CSR) format.
-    CSR
+    CSR,
+    /// Compressed sparse fiber (CSF) format.
```
Just for note: #5932 must be merged before this PR.

mrkn commented on cpp/src/arrow/sparse_tensor.cc:
```diff
@@ -623,6 +795,75 @@ SparseCSRIndex::SparseCSRIndex(const std::shared_ptr<Tensor>& indptr,
+  for (int64_t i = 0; i < ndim; ++i)
+    indices[i] = std::make_shared<Tensor>(indices_type, indices_data[i],
+                                          std::vector<int64_t>({indices_shapes[i]}));
+
```
Why not validate parameters here by `CheckSparseCSFIndexValidity`?

rok replied:
Would it not be checked when the [construcor](https://github.com/apache/arrow/blob/bc0e27831a6b3f4ac27247149a660db7dcc60920/cpp/src/arrow/sparse_tensor.cc#L860) is called?

mrkn replied:
When the validation in the constructor fails, the program is aborted by `ARROW_CHECK`.
To make `Make` function safe, it is necessary to calling the validation in this function.

rok replied:
Got it! Added it now.

mrkn commented on cpp/src/arrow/sparse_tensor.h:
```diff
@@ -228,6 +230,73 @@ class ARROW_EXPORT SparseCSRIndex : public internal::SparseIndexBase<SparseCSRIn
+class ARROW_EXPORT SparseCSFIndex : public internal::SparseIndexBase<SparseCSFIndex> {
+ public:
+  static constexpr SparseTensorFormat::type format_id = SparseTensorFormat::CSF;
+
```
How do you think about adding `kTypeName` static constant member here likewise [this change](https://github.com/apache/arrow/pull/5932/files#diff-b49bb494ab5354bdf56064136a7355a1R291)?

rok replied:
I don't mind, but I also don't see immediate use, but I trust you do :).
I added it like in ARROW-4225.

mrkn commented on cpp/src/arrow/sparse_tensor_test.cc:
```diff
@@ -712,4 +712,194 @@ INSTANTIATE_TYPED_TEST_CASE_P(TestInt64, TestSparseCSRMatrixForIndexValueType, I
+  std::vector<std::shared_ptr<Buffer>> indptr_buffers(3);
+  std::vector<std::shared_ptr<Buffer>> indices_buffers(4);
+  std::vector<int64_t> axis_order = {0, 1, 2, 3};
+  std::vector<int64_t> sparse_tensor_shape({3, 3, 3, 4});
```
This shape doesn't involve to change `axis_order`, so we also need to test another shape like `{4, 3, 5, 2}` to test the `axis_order`-changing case.

rok replied:
Good idea, will work on this.

mrkn commented on cpp/src/arrow/sparse_tensor.cc:
```diff
@@ -301,6 +303,130 @@ class SparseTensorConverter<TYPE, SparseCSRIndex>
+            ->indices();
+
+    // Convert SparseCOOTensor to long CSF buffers
+    const int64_t ndim = tensor_.ndim();
```
How this converter work for the cases of `ndim == 0` and `ndim == 1`?

rok replied:
Added a check for `ndim > 1`.

rok replied:
I'm not sure if and how to address this. I propose we do it when it becomes a problem and for all sparse types at once.

rok commented:
> Thank you very much @rok. Here are some comments.

Thanks for the review @mrkn! I'll through it tonight.

rok commented on cpp/src/arrow/sparse_tensor.cc:
```diff
@@ -623,6 +795,75 @@ SparseCSRIndex::SparseCSRIndex(const std::shared_ptr<Tensor>& indptr,
+                                          const std::vector<int64_t>& indices_shape,
+                                          const int64_t axis_order_size) {
+  if (!is_integer(indptr_type->id())) {
+    return Status::Invalid("Type of SparseCSFIndex indptr must be integer");
```
Done.

rok commented on cpp/src/arrow/sparse_tensor.cc:
```diff
@@ -623,6 +795,75 @@ SparseCSRIndex::SparseCSRIndex(const std::shared_ptr<Tensor>& indptr,
+    return Status::Invalid("Type of SparseCSFIndex indptr must be integer");
+  }
+  if (!is_integer(indices_type->id())) {
+    return Status::Invalid("Type of SparseCSFIndex indices must be integer");
```
Done.

rok commented on cpp/src/arrow/sparse_tensor.cc:
```diff
@@ -623,6 +795,75 @@ SparseCSRIndex::SparseCSRIndex(const std::shared_ptr<Tensor>& indptr,
+  }
+  if (num_indptrs + 1 != num_indices) {
+    return Status::Invalid(
+        "SparseCSFIndex length indices must be equal to length inptrs plus one.");
```
Good point! I'm not a native speaker either.
I proposed a new order.

rok commented on cpp/src/arrow/sparse_tensor.cc:
```diff
@@ -623,6 +795,75 @@ SparseCSRIndex::SparseCSRIndex(const std::shared_ptr<Tensor>& indptr,
+  }
+  if (axis_order_size != num_indices) {
+    return Status::Invalid(
+        "SparseCSFIndex length of indices must be equal number of dimensions.");
```
Proposed a new order.

mrkn commented on cpp/src/arrow/sparse_tensor.cc:
```diff
@@ -422,6 +424,139 @@ class SparseTensorConverter<TYPE, SparseCSCIndex>
+            ->indices();
+
+    // TODO(rok): Coords should be sorted with axis_order priority to improve compression.
+    // ARROW-4221 would help here as well.
```
@rok Why do you need ARROW-4221 (coords' canonical flag) to sort the coords tensor?
In the current implementation, the coords tensor in SparseCOOTensor should always be sorted in the index's lexicographical order.
Doesn't this restriction help you to sort the coords with axis_order?

rok replied:
Changing axis order changes lexicographical order of the COO indices matrix. I've changed the approach now, so this is no longer an issue.

mrkn commented on cpp/src/arrow/sparse_tensor.cc:
```diff
@@ -503,13 +639,40 @@ Status MakeSparseTensorFromTensor(const Tensor& tensor,
 }
 
+template <typename TYPE, typename IndexValueType>
+void assign_values(int64_t dimension, int64_t offset, int64_t first_ptr, int64_t last_ptr,
```
This function should be put in the anonymous namespace.

mrkn replied:
And could you please rename it to more specific name.
For example, since this function expands the sparse tensor into the dense tensor, one of the candidates can be `ExpandSparseCSFTensorValues`.

rok replied:
Done.

mrkn commented on cpp/src/arrow/sparse_tensor.cc:
```diff
@@ -753,6 +936,79 @@ void CheckSparseCSXIndexValidity(const std::shared_ptr<DataType>& indptr_type,
+      indptr_(indptr),
+      indices_(indices),
+      axis_order_(axis_order) {
+  ARROW_CHECK(CheckSparseCSFIndexValidity(indptr_.front()->type(),
```
You can use `ARROW_CHECK_OK` here.

rok replied:
Done! :)

mrkn commented on cpp/src/arrow/sparse_tensor.cc:
```diff
@@ -753,6 +936,79 @@ void CheckSparseCSXIndexValidity(const std::shared_ptr<DataType>& indptr_type,
+    indices[i] = std::make_shared<Tensor>(indices_type, indices_data[i],
+                                          std::vector<int64_t>({indices_shapes[i]}));
+
+  ARROW_CHECK(CheckSparseCSFIndexValidity(indptr_type, indices_type, indptr.size(),
```
Could you please use `RETURN_NOT_OK` instead of `ARROW_CHECK` here because `ARROW_CHECK` aborts the program if the check fails.

mrkn commented on cpp/src/arrow/sparse_tensor_test.cc:
```diff
@@ -903,4 +903,207 @@ TEST_F(TestSparseCSCMatrix, TestToTensor) {
+  std::vector<int64_t> axis_order = {3, 1, 0, 2};
+  std::vector<int64_t> indices_shapes = {3, 6, 6, 8};
+
+  for (int64_t i = 0; i < static_cast<int64_t>(indptr_values.size()); ++i)
```
I guess this for statement can be written by the range-based style:

```cxx
std::vector<std::shared_ptr<Buffer>> indptr_buffers;  // NOTE: remove the count argument
...
for (auto& indptr : indptr_values)
  indptr_buffers.push_back(Buffer::Wrap(indptr));
```

The `indices_buffers` below can be initialized by the same style.

rok replied:
Done.

mrkn commented on format/SparseTensor.fbs:
```diff
@@ -25,7 +25,7 @@ namespace org.apache.arrow.flatbuf;
 /// EXPERIMENTAL: Data structures for sparse tensors
 
-/// Coodinate (COO) format of sparse tensor index.
+/// Coordinate (COO) format of sparse tensor index.
```
Good catch!

mrkn commented on cpp/src/arrow/sparse_tensor_test.cc:
```diff
@@ -903,4 +903,207 @@ TEST_F(TestSparseCSCMatrix, TestToTensor) {
+
+  //  std::shared_ptr<Tensor> dense_tensor;
+  //  ASSERT_OK(sparse_tensor->ToTensor(&dense_tensor));
+  //  ASSERT_TRUE(tensor.Equals(*dense_tensor));
```
Why are these lines commented out?

rok commented:
Thanks for the review @mrkn! I've implemented your feedback.
I've also changed the way `MakeTensorFromSparseTensor` transforms dense to `CSF` (no longer via `COO` tensor) and added more tests.
I'm still working on a type issue in [`TestSparseCSFTensorForIndexValueType, CreateFromBuffers` test](https://github.com/apache/arrow/pull/5716/files#diff-70392006d8af2a80248cf114a7337b0dR1108).

mrkn commented:
Thank you for your work, @rok.  Direct conversion from dense to CSF is better than the conversion via COO.  It's a great improvement, I think.

rok commented:
@mrkn - I've resolved the issue in `MakeTensorFromSparseTensor` so I feel this is ready for review. (The CI error is unrelated;)

rok commented:
@mrkn @pitrou this is ready for a review in case you can spare some time. :)

mrkn commented:
I’ll look this tomorrow.

mrkn commented on format/SparseTensor.fbs:
```diff
@@ -113,9 +113,71 @@ table SparseMatrixIndexCSX {
+  ///
+  ///   indptrBuffer(X) = [[0, 2, 3], [0, 1, 3, 4], [0, 2, 4, 5, 8]].
+  ///
+  indptrBuffers: [Buffer];
```
Could you please mark `indptrBuffers`, `indicesBuffers`, and `axisOrder` property required?
You can see how to do this in https://github.com/apache/arrow/pull/6234.

rok replied:
Done. I've also made `indptrType` and `indicesType` required since I see `SparseMatrixIndexCSX` and `SparseTensorIndexCOO` have it as required.

pitrou commented on cpp/src/arrow/sparse_tensor_test.cc:
```diff
@@ -910,4 +911,326 @@ TEST_F(TestSparseCSCMatrix, TestToTensor) {
+      SparseCSFIndex::Make(TypeTraits<IndexValueType>::type_singleton(), indices_shapes,
+                           axis_order, indptr_buffers, indices_buffers));
+  std::shared_ptr<SparseCSFTensor> st =
+      std::make_shared<SparseCSFTensor>(si, int64(), data_buffer, shape, dim_names);
```
This code is used several time accross the tests. Can you factor it out, please?

rok replied:
Done.

pitrou commented on cpp/src/arrow/sparse_tensor_test.cc:
```diff
@@ -910,4 +911,326 @@ TEST_F(TestSparseCSCMatrix, TestToTensor) {
+      0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
+      0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
+      1, 2, 0, 0, 0, 0, 0, 0, 0, 0, 0, 3, 0, 4, 0, 5, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
+      0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 6, 7, 8};
```
At the minimum, you should factor out those examples, e.g. in a helper function or method (and/or write them using a more readable idiom as suggested above).

rok replied:
I refactored a bit. Please check if it's dry enough.

pitrou commented on cpp/src/arrow/sparse_tensor_test.cc:
```diff
@@ -910,4 +911,326 @@ TEST_F(TestSparseCSCMatrix, TestToTensor) {
+      1, 2, 0, 0, 0, 0, 0, 0, 0, 0, 0, 3, 0, 4, 0, 5, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
+      0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 6, 7, 8};
+
+  std::vector<int64_t> dense_values_2 = {
```
Please find another idiom for this.

pitrou commented on cpp/src/arrow/sparse_tensor.cc:
```diff
@@ -419,6 +421,152 @@ class SparseTensorConverter<TYPE, SparseCSCIndex>
+    RETURN_NOT_OK(CheckMaximumValue(std::numeric_limits<c_index_value_type>::max()));
+
+    const int64_t ndim = tensor_.ndim();
+    std::vector<int64_t> axis_order = internal::ArgSort(tensor_.shape());
```
Add a comment explaining the heuristic?

pitrou commented on cpp/src/arrow/sparse_tensor.cc:
```diff
@@ -419,6 +421,152 @@ class SparseTensorConverter<TYPE, SparseCSCIndex>
+      for (int64_t n = tensor_.size(); n > 0; n--) {
+        const value_type x = tensor_.Value(coord);
+
+        if (tensor_.Value(coord) != 0) {
```
`if (x != 0)`

pitrou commented on cpp/src/arrow/sparse_tensor.cc:
```diff
@@ -419,6 +421,152 @@ class SparseTensorConverter<TYPE, SparseCSCIndex>
+            if (tree_split || change) {
+              if (change) tree_split = true;
+
+              if (i < ndim - 1)
```
Please use braces after `if`, to avoid mistakes.

pitrou commented on cpp/src/arrow/sparse_tensor_test.cc:
```diff
@@ -910,4 +911,326 @@ TEST_F(TestSparseCSCMatrix, TestToTensor) {
+}
+
+TYPED_TEST_P(TestSparseCSFTensorForIndexValueType, TestSparseTensorToTensor) {
+  std::vector<int64_t> shape = {3, 3, 3, 4};
```
Can you add a test with a non-ascending shape?

rok replied:
Added `TestNonAscendingShape`.

pitrou commented on cpp/src/arrow/sparse_tensor.cc:
```diff
@@ -419,6 +421,152 @@ class SparseTensorConverter<TYPE, SparseCSCIndex>
+          bool tree_split = false;
+          *values++ = x;
+
+          for (int64_t i = 0; i < ndim; ++i) {
```
Hmm, you are walking in logical dimension order. Shouldn't you walk in physical dimension order instead?

Your dense-to-sparse tests all seem to use an ascending shape, so of course `dimension == i` in that case...

rok replied:
We need to walk it in logical order as structure of the graph changes depending on the order of elements walked. (By physical I mean the way the dense tensor is layed out in memory and by logical the way it would have been if dimensions were argsorted.)

The code was wrong indeed - `counts` should have been updated in physical order and `coords` was incremented in physical not logical order.

I've added `TestNonAscendingShape` to test.

pitrou commented on cpp/src/arrow/sparse_tensor.cc:
```diff
@@ -419,6 +421,152 @@ class SparseTensorConverter<TYPE, SparseCSCIndex>
+          previous_coord = coord;
+        }
+
+        // increment index
```
Don't we have similar code (incrementing a multi-dimensional index) elsewhere in this file? If reasonably doable, it should be factored out.

rok replied:
There is similar code for `COO` sparse tensor but for `CSF` we need to use `axis_order` to correctly increment. We could factor this out but then we need to have `COO` create and pass `axis_order`. I'm not sure that would improve the situation.

mrkn replied:
Related to this, I considered we may need a multi-dimensional index iterator for tensor and sparse tensor before. But I've not added it yet because I've never needed it in the current implementation.

How do you think about extracting index incrementing routines as two functions with the same name: one takes `axis_order`, and the other is specialized for without `axis_order`?

rok replied:
I've factored out `IncrementIndex` functions for CSF and COO. Please check if it makes sense. Should we do this for CSR/CSC too?

pitrou commented on cpp/src/arrow/sparse_tensor.cc:
```diff
@@ -419,6 +421,152 @@ class SparseTensorConverter<TYPE, SparseCSCIndex>
+    std::vector<int64_t> indptr_shapes(counts.begin(), counts.end() - 1);
+    std::vector<int64_t> indices_shapes = counts;
+
+    for (int64_t column = 0; column < ndim; ++column)
```
Please use braces after `for`.

rok commented on format/SparseTensor.fbs:
```diff
@@ -114,9 +114,71 @@ table SparseMatrixIndexCSX {
+  /// Position in the indptrBuffers vector signifies the dimension.
+  /// For example, the indptrBuffers for the above X is:
+  ///
+  ///   indptrBuffer(X) = [[0, 2, 3], [0, 1, 3, 4], [0, 2, 4, 5, 8]].
```
My original comment was a bit off. I rephrased it, please check.

rok commented on cpp/src/arrow/sparse_tensor_test.cc:
```diff
@@ -910,4 +911,326 @@ TEST_F(TestSparseCSCMatrix, TestToTensor) {
+    //   X[2, 2, 2, 2] := 7
+    //   X[2, 2, 2, 3] := 8
+
+    std::vector<int64_t> dense_values = {
```
Nice idea for initialization. I've also switched to `int16_t` and refactored a bit.
Please see if you find this dry enough.

rok commented:
Thanks for an extensive review @pitrou!
I've addressed the points, please review. :)

> * I notice that IPC is not tested though you added the Flatbuffers definitions. If it's for a separate PR, please open a JIRA for it :-)

I've opened [ARROW-7428](https://issues.apache.org/jira/browse/ARROW-7428) and will work on it next. Would it be better to move the flatbuffers definition to the new PR?

pitrou commented:
> I've opened ARROW-7428 and will work on it next. Would it be better to move the flatbuffers definition to the new PR?

That sounds reasonable, yes.

rok commented:
> > I've opened ARROW-7428 and will work on it next. Would it be better to move the flatbuffers definition to the new PR?

> That sounds reasonable, yes.

I've opened #6340 and moved flatbuffers code there.

mrkn commented on cpp/src/arrow/sparse_tensor.cc:
```diff
@@ -55,6 +57,36 @@ class SparseTensorConverter {
+// IncrementIndex for SparseCOOIndex and SparseCSFIndex
+
+void IncrementIndex(std::vector<int64_t>& coord, int64_t n, std::vector<int64_t> shape,
+                    int64_t ndim) {
```
The type of `shape` should be `const std::vector<int64_t>&`, and `ndim` should be taken from `shape.size()`.

mrkn replied:
It is better to make `n` const because `n` isn't changed in this function.

rok replied:
Good point! Refactoring. Sorry I missed this.

mrkn commented on cpp/src/arrow/sparse_tensor.cc:
```diff
@@ -55,6 +57,36 @@ class SparseTensorConverter {
+}
+
+void IncrementIndex(std::vector<int64_t>& coord, int64_t n, std::vector<int64_t> shape,
+                    int64_t ndim, std::vector<int64_t> axis_order) {
```
Same as above.

mrkn commented on cpp/src/arrow/sparse_tensor.cc:
```diff
@@ -128,17 +160,7 @@ class SparseTensorConverter<TYPE, SparseCOOIndex>
-            --d;
-          }
-        }
+        IncrementIndex(coord, n, shape, ndim);
```
I guess that `n` can be removed from the parameters of `IncrementIndex` by rewriting as follow.

```
if (n > 1) {
  IncrementIndex(coord, shape, ndim);
}
```

mrkn commented on cpp/src/arrow/sparse_tensor.cc:
```diff
@@ -419,6 +441,146 @@ class SparseTensorConverter<TYPE, SparseCSCIndex>
+          }
+          previous_coord = coord;
+        }
+        IncrementIndex(coord, n, shape, ndim, axis_order);
```
I guess `n` can be removed here in the same way.

pitrou commented on cpp/src/arrow/sparse_tensor.cc:
```diff
@@ -419,6 +444,148 @@ class SparseTensorConverter<TYPE, SparseCSCIndex>
+            bool change = coord[dimension] != previous_coord[dimension];
+
+            if (tree_split || change) {
+              if (change) tree_split = true;
```
Can this be simplified a bit thusly?
```c++
tree_split = tree_split || (coord[dimension] != previous_coord[dimension]);
if (tree_split) {
```

rok replied:
Nice idea!

pitrou commented on cpp/src/arrow/sparse_tensor.cc:
```diff
@@ -419,6 +444,148 @@ class SparseTensorConverter<TYPE, SparseCSCIndex>
+          }
+          previous_coord = coord;
+        }
+        if (n > 1) {
```
Same remark as above.

pitrou commented on cpp/src/arrow/sparse_tensor.cc:
```diff
@@ -500,13 +668,45 @@ Status MakeSparseTensorFromTensor(const Tensor& tensor,
+void ExpandSparseCSFTensorValues(int64_t dimension, int64_t offset, int64_t first_ptr,
+                                 int64_t last_ptr, const SparseCSFIndex* sparse_index,
+                                 const TYPE* raw_data, const std::vector<int64_t> strides,
+                                 const std::vector<int64_t> axis_order, TYPE* out) {
```
Pass vectors as `const&`.

rok replied:
Got it. Would it make sense to make `int64_t`s `const` too?

pitrou replied:
No, they are just values.

pitrou commented on cpp/src/arrow/sparse_tensor.cc:
```diff
@@ -500,13 +668,45 @@ Status MakeSparseTensorFromTensor(const Tensor& tensor,
+
+template <typename TYPE, typename IndexValueType>
+void ExpandSparseCSFTensorValues(int64_t dimension, int64_t offset, int64_t first_ptr,
+                                 int64_t last_ptr, const SparseCSFIndex* sparse_index,
```
`const SparseCSFIndex&`?

pitrou commented on cpp/src/arrow/sparse_tensor.cc:
```diff
@@ -500,13 +668,45 @@ Status MakeSparseTensorFromTensor(const Tensor& tensor,
+namespace {
+
+template <typename TYPE, typename IndexValueType>
+void ExpandSparseCSFTensorValues(int64_t dimension, int64_t offset, int64_t first_ptr,
```
Rename `offset` to `dense_offset`?

pitrou commented on cpp/src/arrow/sparse_tensor.cc:
```diff
@@ -754,6 +975,76 @@ void CheckSparseCSXIndexValidity(const std::shared_ptr<DataType>& indptr_type,
+  }
+  if (num_indptrs + 1 != num_indices) {
+    return Status::Invalid(
+        "Length of indices must be equal to length of inptrs + 1 for SparseCSFIndex.");
```
Nit : "indptrs + 1"

pitrou commented on cpp/src/arrow/sparse_tensor.cc:
```diff
@@ -754,6 +975,76 @@ void CheckSparseCSXIndexValidity(const std::shared_ptr<DataType>& indptr_type,
+  }
+  if (axis_order_size != num_indices) {
+    return Status::Invalid(
+        "Length of indices must be equal number of dimensions for SparseCSFIndex.");
```
Nit: "equal to"

pitrou commented on cpp/src/arrow/sparse_tensor.h:
```diff
@@ -329,6 +331,74 @@ class ARROW_EXPORT SparseCSCIndex
+  std::string ToString() const override;
+
+  /// \brief Return whether the CSF indices are equal
+  bool Equals(const SparseCSFIndex& other) const {
```
I don't this needs to be inlined in the `.h` file.
Also, please use braces with `if`s and `for`s.

rok replied:
Moved to `.cc`.

pitrou commented on cpp/src/arrow/sparse_tensor.h:
```diff
@@ -329,6 +331,74 @@ class ARROW_EXPORT SparseCSCIndex
+                          std::vector<std::shared_ptr<Tensor>>& indices,
+                          const std::vector<int64_t>& axis_order);
+
+  /// \brief Return a 1D tensor of indptr vector
```
"a vector of 1D indptr tensors"?

pitrou commented on cpp/src/arrow/sparse_tensor.h:
```diff
@@ -329,6 +331,74 @@ class ARROW_EXPORT SparseCSCIndex
+  /// \brief Return a 1D tensor of indptr vector
+  const std::vector<std::shared_ptr<Tensor>>& indptr() const { return indptr_; }
+
+  /// \brief Return a 1D tensor of indices vector
```
"a vector of 1D indices tensors"?

pitrou commented on cpp/src/arrow/sparse_tensor_test.cc:
```diff
@@ -26,6 +26,7 @@
 
 #include <gtest/gtest.h>
 
+#include <arrow/util/sort.h>
```
Please leave includes in alphabetical order, and use quotes here: `#include "arrow/util/sort.h"`

pitrou commented on cpp/src/arrow/sparse_tensor_test.cc:
```diff
@@ -910,4 +911,242 @@ TEST_F(TestSparseCSCMatrix, TestToTensor) {
+ protected:
+  std::shared_ptr<SparseCSFIndex> MakeSparseCSFIndex(
+      const std::vector<int64_t> axis_order,
+      std::vector<std::vector<typename IndexValueType::c_type>>& indptr_values,
```
`const&` for const vectors.

pitrou replied:
You don't seem to mutate those vector inputs, so should make them a const reference really :-)

pitrou commented on cpp/src/arrow/sparse_tensor_test.cc:
```diff
@@ -910,4 +911,242 @@ TEST_F(TestSparseCSCMatrix, TestToTensor) {
+  template <typename CValueType>
+  std::shared_ptr<SparseCSFTensor> MakeSparseTensor(
+      const std::shared_ptr<SparseCSFIndex>& si, std::vector<CValueType>& sparse_values,
+      const std::vector<int64_t> shape, const std::vector<std::string> dim_names) const {
```
`const&` for vectors

pitrou commented on cpp/src/arrow/sparse_tensor_test.cc:
```diff
@@ -910,4 +911,242 @@ TEST_F(TestSparseCSCMatrix, TestToTensor) {
+  std::shared_ptr<SparseCSFTensor> sparse_tensor;
+  ASSERT_OK_AND_ASSIGN(
+      sparse_tensor,
+      SparseCSFTensor::Make(dense_tensor, TypeTraits<IndexValueType>::type_singleton()));
```
It seems this is doing the same thing as the `SetUp` function. Why not test `sparse_tensor_from_dense_` directly?

rok replied:
Good point. Fixed.

mrkn commented on cpp/src/arrow/sparse_tensor.cc:
```diff
@@ -55,6 +57,37 @@ class SparseTensorConverter {
+// ----------------------------------------------------------------------
+// IncrementIndex for SparseCOOIndex and SparseCSFIndex
+
+void IncrementIndex(std::vector<int64_t>& coord, const std::vector<int64_t> shape) {
```
Yes, `shape` should be `const&`, and this function should be marked `inline`.

rok replied:
Done.

rok commented on cpp/src/arrow/python/serialize.cc:
```diff
@@ -665,12 +666,15 @@ Status CountSparseTensors(
         break;
       case SparseTensorFormat::CSC:
         // TODO(mrkn): support csc
```
Fixed.

rok commented:
Thanks for the review @pitrou! I've pushed changes you suggested.
It's good to see this is slowly coming together.

rok commented:
It seems [flight tests are failing](https://ci.appveyor.com/project/ApacheSoftwareFoundation/arrow/builds/30573849/job/04q4c4wp2i36ua6g#L2218) in appveyor.

pitrou commented on cpp/src/arrow/sparse_tensor.cc:
```diff
@@ -754,6 +971,86 @@ void CheckSparseCSXIndexValidity(const std::shared_ptr<DataType>& indptr_type,
+    const std::shared_ptr<DataType>& indices_type,
+    const std::vector<int64_t>& indices_shapes, const std::vector<int64_t>& axis_order,
+    std::vector<std::shared_ptr<Buffer>> indptr_data,
+    std::vector<std::shared_ptr<Buffer>> indices_data) {
```
`const vector&` here too. `vector`s are expensive to copy, so unless the function is mutating the vector input, it should take it as a const reference.

rok replied:
Fixed! Thanks for the patience :).

pitrou commented on cpp/src/arrow/sparse_tensor.cc:
```diff
@@ -754,6 +971,86 @@ void CheckSparseCSXIndexValidity(const std::shared_ptr<DataType>& indptr_type,
+
+// Constructor with two index vectors
+SparseCSFIndex::SparseCSFIndex(std::vector<std::shared_ptr<Tensor>>& indptr,
+                               std::vector<std::shared_ptr<Tensor>>& indices,
```
Unless you plan to mutate those vectors, you should pass them as const-reference (otherwise the caller may assume you're going to mutate them).

pitrou commented on cpp/src/arrow/sparse_tensor.h:
```diff
@@ -329,6 +331,68 @@ class ARROW_EXPORT SparseCSCIndex
+      const std::shared_ptr<DataType>& indices_type,
+      const std::vector<int64_t>& indices_shapes, const std::vector<int64_t>& axis_order,
+      std::vector<std::shared_ptr<Buffer>> indptr_data,
+      std::vector<std::shared_ptr<Buffer>> indices_data) {
```
const references here too.

rok commented:
Thanks for the review @pitrou. I pushed the suggested changes :).

pitrou commented:
Thank you @rok :-)

rok commented:
Thanks @pitrou @mrkn and @ShadenSmith!
